cfgen/cfgen

```diff
@@ -831,22 +843,36 @@ def oids_to_dhall(oids: List[Oid]) -> str:
 NetProtocol = Enum('NetProtocol', 'tcp o2ib')
 
 
-ProcT = Enum('ProcT', 'hax m0_server m0_client_s3 m0_client_other')
+ProcT = Enum('ProcT', 'hax m0_ios m0_client_s3 m0_client_other')
 ProcT.__doc__ = 'Type of process'
 
 
-class LnetPortalGroup(IntEnum):
-    hax = 1,
-    m0_server = 2,
-    m0_client_s3 = 3,
-    m0_client_other = 4
+class LnetPortalGroup:
+    def __init__(self, hax=1, m0_ios=2, m0_client_s3=3,
+                 m0_client_other=4):
+        self.hax = hax
+        self.m0_ios = m0_ios
+        self.m0_client_s3 = m0_client_s3
+        self.m0_client_other = m0_client_other
+
+    def port_next(self, portal_group_name: str) -> int:
+        setattr(self, portal_group_name,
+                (getattr(self, portal_group_name) + 1))
+        return int(getattr(self, portal_group_name))
 
 
-class LibfabPortalGroup(IntEnum):
-    hax = 2000,
-    m0_server = 3000,
-    m0_client_s3 = 4000,
-    m0_client_other = 5000
+class LibfabPortalGroup:
```

Comment:
@rahul27kumar, sorry, I think we can maintain the much of the earlier implementation by allocating the enum dynamically, e.g.
```
class LibfabPortalGroup:
    def __init__(self, hax=22000, m0_ios=21000, m0_client_s3=22500, m0_client_other=21500):
        self.portalgroup = enum.Enum('LibfabPortalGroup', {'hax':hax, 'm0_ios':m0_ios, ...})

# Initialize LibfabPortalGroup
LibfabPortalGroup(hax=22000, ...)

# use LibfabPortalGroup
class LibfabricEndpoint:
    def __init__(self, NetFamily: str, proto: NetProtocol, ipaddr: str, portalgroup: LibfabPortalGroup) ===> don't need proct

LibfabricEndpoint(netfamily, ipaddr, LibfabPortalGroup.portalgroup[proct],..)

Most of the implementation will remain the same.

Reply:
@rahul27kumar, I think we can implement the above as follows to make mypy happy,
```
class LibfabPortalGroup:
    def __init__(self, hax=21000, m0server=22000, ...):
        self.hax = hax
        self.m0server = m0server
        ...

    def get_portal_group(self) -> Enum:
        return Enum('LibfabPortalGroup', {'hax': self.hax,
                                      'm0server': self.m0server})
```
Then while instantiating the PortalGroup we do (at line #1131)
```
portalgroup=LibfabPortalGroup(hax=ports['hax'], ...).get_portal_group()[proc_t.name]
```
can you try this, build completed successfully for me.

Reply:
Thank you @mssawant Let me try

Reply:
@mssawant 
``` 
(.py3venv) [root@ssc-vm-g4-rhev4-0086 cfgen]# make
flake8  cfgen
set -eu -o pipefail; for f in cfgen; do MYPYPATH=../stubs:../hax mypy --config-file ../hax/mypy.ini  $f; done
cfgen:1121: error: Value of type "Enum" is not indexable
Found 1 error in 1 file (checked 1 source file)
make: *** [typecheck] Error 1
```

portalgroup=LibfabPortalGroup(**ports).get_portal_group()[proc_t.name]
or
portalgroup=LibfabPortalGroup(hax=ports['hax],...).get_portal_group()[proc_t.name]

Tried above approaches, but it gives same error message
 CC @vaibhavparatwar

Reply:
Looks like mypy is not able to figure out, we can workaround mypy for this situation simply by avoiding return type hint as

```
index 8c94de1..4d22ca0 100755
--- a/cfgen/cfgen
+++ b/cfgen/cfgen
@@ -3,7 +3,6 @@
import abc
import argparse
import ast
-import enum
from enum import Enum, auto, IntEnum
from functools import lru_cache, wraps

```
import itertools
@@ -852,9 +851,12 @@ class LibfabPortalGroup(IntEnum):



class LibfabPortals:
def __init__(self, hax=21000, m0server=22000):
- self.portals = enum.Enum('LibfabPortals',
- {'hax': hax,
- 'm0server': m0server})
+ self.hax = hax
+ self.m0server = m0server
+
+ def get_portal_group(self):
+ return Enum('LibfabPortals', {'hax': self.hax,
+ 'm0server': self.m0server})




class Endpoint:
@@ -1102,11 +1104,12 @@ class ConfProcess(ToDhall):
facts = node_desc['facts']
proto = node_desc.get('data_iface_type', 'tcp')
transport_type = node_desc.get('transport_type', 'libfab')
+ portals = LibfabPortals(hax=23000, m0server=24000).get_portal_group()
ep: Any = (
LibfabricEndpoint(NetFamily='inet', proto=NetProtocol[proto],
ipaddr=facts[ipaddr_key(
node_desc['data_iface'])],
- portalgroup=LibfabPortalGroup[proc_t.name]))
+ portalgroup=portals[proc_t.name]))
if transport_type == 'lnet':
# Creates Lnet endpoint, remove once deprecated.
ep = LnetEndpoint(proto=NetProtocol[proto],
```
can you try this?